src/tune-hyperparameters/src/main/scala/TuneHyperparameters.scala

```diff
@@ -0,0 +1,221 @@
+    */
+  private def awaitResult[T](awaitable: Awaitable[T], atMost: Duration): T = {
+    try {
+      val awaitPermission = null.asInstanceOf[scala.concurrent.CanAwait]
+      awaitable.result(atMost)(awaitPermission)
+    } catch {
+      case NonFatal(t) if !t.isInstanceOf[TimeoutException] =>
+        throw new SparkException("Exception thrown in awaitResult: ", t)
+    }
+  }
+
+  /** Finds the best model
+    *
+    * @param dataset The input dataset to train.
+    * @return The trained classification model.
+    */
+  override def fit(dataset: Dataset[_]): TuneHyperparametersModel = {
+    val sparkSession = dataset.sparkSession
+    val splits = MLUtils.kFold(dataset.toDF.rdd, getNumFolds, getSeed)
+    val hyperParams = getParamSpace.paramMaps
+    val schema = dataset.schema
+    val executionContext = getExecutionContext
+    val (evaluationMetricColumnName, operator): (String, Ordering[Double]) =
+      EvaluationUtils.getMetricWithOperator(getModels.head, getEvaluationMetric)
+    val paramsPerRun = ListBuffer[ParamMap]()
+    for (_ <- 0 until getNumRuns) {
+      // Generate the new parameters, stepping through estimators sequentially
+      paramsPerRun += hyperParams.next()
+    }
+    val numModels = getModels.length
+
+    val metrics = splits.zipWithIndex.map { case ((training, validation), splitIndex) =>
+      val trainingDataset = sparkSession.createDataFrame(training, schema).cache()
+      val validationDataset = sparkSession.createDataFrame(validation, schema).cache()
+
+      val modelFutures = ListBuffer[Future[Model[_]]]()
+      val modelFuturesParams = ListBuffer[ParamMap]()
+      for (n <- 0 until getNumRuns) {
+        val params = paramsPerRun(n)
+        modelFuturesParams += params
+        modelFutures += Future[Model[_]] {
+          val model = getModels(n % numModels).fit(trainingDataset, params)
+          model.asInstanceOf[Model[_]]
+        }
+      }
+
+      // Unpersist training data only when all models have trained
+      Future.sequence[Model[_], Iterable](modelFutures)(implicitly, executionContext)
+        .onComplete { _ => trainingDataset.unpersist() } (executionContext)
+
+      // Evaluate the models
+      val foldMetricFutures = modelFutures.zip(modelFuturesParams).map { case (modelFuture, paramMap) =>
+        modelFuture.map { model =>
+          val scoredDataset = model.transform(validationDataset, paramMap)
+          val evaluator = new ComputeModelStatistics()
+          evaluator.set(evaluator.evaluationMetric, getEvaluationMetric)
+          model match {
+            case _: TrainedRegressorModel => {
+              logDebug("Evaluating trained regressor model.")
+            }
+            case _: TrainedClassifierModel => {
+              logDebug("Evaluating trained classifier model.")
+            }
+            case classificationModel: ClassificationModel[_, _] => {
+              logDebug(s"Evaluating SparkML ${model.uid} classification model.")
```

Comment:
This kind of structure is what worries me, there are very large case matches throughout, which points to the need to re-abstract some of this

Reply:
how can I improve the case matches?  It seems like it is necessary in some cases.